lib/manager/terraform/lockfile/index.ts

```diff
@@ -0,0 +1,106 @@
+import pMap from 'p-map';
+import { GetPkgReleasesConfig, getPkgReleases } from '../../../datasource';
+import { logger } from '../../../logger';
+import { get as getVersioning } from '../../../versioning';
+import type {
+  UpdateArtifact,
+  UpdateArtifactsConfig,
+  UpdateArtifactsResult,
+} from '../../types';
+import hash from './hash';
+import type { ProviderLock, ProviderLockUpdate } from './types';
+import {
+  extractLocks,
+  isPinnedVersion,
+  readLockFile,
+  writeLockUpdates,
+} from './util';
+
+async function updateAllLocks(
+  locks: ProviderLock[],
+  config: UpdateArtifactsConfig
+): Promise<ProviderLockUpdate[]> {
+  const updates = await pMap(
+    locks,
+    async (lock) => {
+      const updateConfig: GetPkgReleasesConfig = {
+        versioning: 'hashicorp',
+        datasource: 'terraform-provider',
+        depName: lock.lookupName,
+      };
+      const releasesResult = await getPkgReleases(updateConfig);
+      const releases = releasesResult.releases;
+      const versioning = getVersioning(updateConfig.versioning);
+      const versionsList = releases.map((release) => release.version);
+      const newVersion = versioning.getSatisfyingVersion(
+        versionsList,
+        lock.constraints
+      );
+
+      // if the new version is the same as the last, signal that no update is needed
+      if (newVersion === lock.version) {
+        return null;
+      }
+      const update: ProviderLockUpdate = {
+        newVersion,
+        newConstraint: lock.constraints,
+        newHashes: await hash(lock.lookupName, newVersion, config.cacheDir),
+        ...lock,
+      };
+      return update;
+    },
+    { concurrency: 4 } // allow to look up 4 lock in parallel
+  );
+
+  return updates.filter(Boolean);
+}
+
+export async function updateArtifacts({
+  packageFileName,
+  updatedDeps,
+  newPackageFileContent,
+  config,
+}: UpdateArtifact): Promise<UpdateArtifactsResult[] | null> {
+  logger.debug(`terraform.updateArtifacts(${packageFileName})`);
+
+  const lockFileContent = await readLockFile();
```

Comment:
We can use `fs.findLocalSiblingOrParent` later to search upwards from `packageFileName`, but maybe `fs.getSiblingFileName` would be enough as the lock file should be in same folder.

https://github.com/renovatebot/renovate/blob/460bfe95f030fcc1f0573d1bc17bb054c5b03cc5/lib/util/fs/index.ts#L108-L117

Reply:
I'm if this would work in mono repos, but I think we should ship it like this and gather feedback first.

Reply:
If it's you ^^ ...